crates/chain/src/keychain/txout_index.rs

```diff
@@ -462,66 +408,82 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
         &self,
     ) -> impl DoubleEndedIterator<Item = (&K, &Descriptor<DescriptorPublicKey>)> + ExactSizeIterator + '_
     {
-        self.keychains_to_descriptors
-            .iter()
-            .map(|(k, (_, d))| (k, d))
+        self.keychains_to_descriptor_ids.iter().map(|(k, did)| {
+            (
+                k,
+                self.descriptor_ids_to_descriptors
+                    .get(did)
+                    .expect("invariant"),
+            )
+        })
     }
 
     /// Insert a descriptor with a keychain associated to it.
     ///
     /// Adding a descriptor means you will be able to derive new script pubkeys under it
     /// and the txout index will discover transaction outputs with those script pubkeys.
     ///
-    /// When trying to add a keychain that already existed under a different descriptor, or a descriptor
-    /// that already existed with a different keychain, the old keychain (or descriptor) will be
-    /// overwritten.
+    /// keychain <-> descriptor is a one-to-one mapping -- in `--release` this method will ignore calls that try to
+    /// associate a keychain with the descriptor of another keychain or to re-assign the keychain to
+    /// new descriptor. If `debug_assertions` are enabled then it will panic.
```

Comment:
🔧 This code now throws an error and should not panic, docs need to be updated to reflect new behavior.
```suggestion
    /// The keychain <-> descriptor relationship is a one-to-one mapping that can not be changed. This method 
    /// will throw an error if you try to associate a new keychain with an already assigned descriptor or to re-assign 
    /// an existing keychain to a new descriptor.
```